[addons] Bring in new binary style (First part)

AlwinEsch opened the pull request:
## Description
This two commits bring in the first parts to allow multiple instances on Kodi.

The first add the new functions in Kodi to create the interface. This are currently Independent from old calls to have the old still compatible.

There becomes now additionally to ADDON_Create the ADDON_CreateInstance used. Due to the "C++" Headers are most parts be done from the classes.

The Second commit change the Screensaver Interface to the new way, who is already used on Kodi agile.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented:
As note, except screensaver, are on no addons changes on code or version needed.

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -0,0 +1,439 @@
+
+  virtual ~CAddonBase()
+  {
+  }
```
<code>virtual ~CAddonBase() = default;</code> maybe?

AlwinEsch replied:
Here and above is changed to the default, it works too.

The destructors are added there to destruct also addons own class. During tests I has seen that without them a destructor on addon itself was not called (don't know why).

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h:
```diff
@@ -0,0 +1,439 @@
+{
+public:
+  IAddonInstance(ADDON_TYPE type) : m_type(type) { }
+  virtual ~IAddonInstance() { }
```
<code>virtual ~IAddonInstance() = default;</code> maybe?

ksooo reviewed (approved):
Did a very careful review, found nothing suspicious. The only thing I want to suggest is to do a cleanup of some log strings and code comments wrt. correct English, but this can be done in a followup PR.

ksooo commented:
jenkins build this please

AlwinEsch commented:
Thanks @ksooo, the english correction can be easily done during add of addon documentation.

AlwinEsch commented:
Also some screensaver addons need to have settings access, due to request https://github.com/xbmc/xbmc/pull/12125 I'm not sure who comes first.

For me the best is to insert https://github.com/xbmc/xbmc/pull/12125 before, that I can change settings functions to new style.

After them is in becomes a commit added here to have on new way.

ksooo commented:
To not delay your work I would opt for merging your stuff first. We can adapt screensaver later.

ksooo commented:
Alternatively, you leave out the screensaver commit and merge only the core change with this PR.

Paxxi commented on xbmc/addons/AddonDll.h:
```diff
@@ -54,19 +76,51 @@ namespace ADDON
+
+    /// addon to kodi basic callbacks below
+    //@{
+    AddonGlobalInterface m_interface;
```
Where does this come from? Tried to find any reference to it and failed.

AlwinEsch replied:
Is here https://github.com/AlwinEsch/kodi/blob/12213e445eb79ab69e8f291dc0b4d2e08bd234f9/xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h#L191-L216 because of the size
 is it not direct visible.

I add a note here to know direct the source

Paxxi commented on xbmc/addons/AddonDll.h:
```diff
@@ -46,6 +46,28 @@ namespace ADDON
+
   protected:
     bool Initialized() { return m_initialized; }
     static uint32_t GetChildCount() { static uint32_t childCounter = 0; return childCounter++; }
```
this isn't threadsafe, somewhat weird design to initialize on a get call and doesn't follow coding style.

Not part of this PR but yeah,

AlwinEsch replied:
Does go away complete soon :D

Paxxi replied:
if it's temporary then it's ok.

AlwinEsch replied:
Is a remnant of the old path of Hack was introduced to make multiple instances of an addon.

There was the addon dll copied to another name to open them independent again.

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -46,6 +48,7 @@ CAddonDll::CAddonDll(AddonProps props)
   m_pHelpers    = NULL;
   m_needsavedsettings = false;
   m_parentLib.clear();
+  memset(&m_interface, 0, sizeof(m_interface));
```
this is generally not a good way to initialize a struct/class, it'll destroy the vtable if it happens to have virtual methods.

AlwinEsch replied:
Is only a poor "C" part. Does a `m_interface = 0` also work?

AlwinEsch replied:
Tried it, a memset with a `= 0` not possible. Have no alternative to them, the structure must complete in "C" style and also not want to add a "C++" constructor there.

Paxxi replied:
`m_interface = {0}` should work on all platforms.

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -236,6 +240,62 @@ ADDON_STATUS CAddonDll::Create(ADDON_TYPE type, void* funcTable, void* info)
+  else if ((status == ADDON_STATUS_NEED_SETTINGS) || (status == ADDON_STATUS_NEED_SAVEDSETTINGS))
+  {
+    m_needsavedsettings = (status == ADDON_STATUS_NEED_SAVEDSETTINGS);
+    if ((status = TransferSettings()) == ADDON_STATUS_OK)
```
don't assign in conditional statements, it saves a line but so easy to mess up and break things.

AlwinEsch replied:
This is currently a c&p from old, this way with "ADDON_STATUS_NEED_SAVEDSETTINGS" becomes complete removed in future.

Wanted to keep the changes as small as possible.

No problem I add a new line.

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -236,6 +240,62 @@ ADDON_STATUS CAddonDll::Create(ADDON_TYPE type, void* funcTable, void* info)
+    if ((status = TransferSettings()) == ADDON_STATUS_OK)
+      m_initialized = true;
+    else
+      new CAddonStatusHandler(ID(), status, "", false);
```
this leaks the status handler

AlwinEsch replied:
The next part who must be removed asap, also a remnant from the old system who comes with c&p to there.

If acceptable for you becomes this done on soon coming requests. Is related to my comment before (only be used for this "ADDON_STATUS_NEED_SAVEDSETTINGS" and "ADDON_STATUS_NEED_SETTINGS")

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -236,6 +240,62 @@ ADDON_STATUS CAddonDll::Create(ADDON_TYPE type, void* funcTable, void* info)
+    CLog::Log(LOGERROR, "ADDON: Dll %s - Client returned bad status (%i) from Create and is not usable", Name().c_str(), status);
+
+    // @todo currently a copy and paste from other function and becomes improved.
+    CGUIDialogOK* pDialog = g_windowManager.GetWindow<CGUIDialogOK>(WINDOW_DIALOG_OK);
```
This should be removed, but currently there's no support for the dialogok through appmessenger so it's ok for now.

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -278,6 +340,42 @@ void CAddonDll::Destroy()
+
+void CAddonDll::DestroyInstance(const std::string& instanceID)
+{
+  std::map<std::string, std::pair<ADDON_TYPE, KODI_HANDLE>>::iterator it = m_usedInstances.find(instanceID);
```
use `auto`

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -425,5 +523,233 @@ bool CAddonDll::CheckAPIVersion(int type)
+ 
+bool CAddonDll::InitInterface(KODI_HANDLE firstKodiInstance)
+{
+  memset(&m_interface, 0, sizeof(m_interface));
```
same as the memset comment above

AlwinEsch replied:
Is removed on this file and also removed them on Screensaver class who was brought before.

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -425,5 +523,233 @@ bool CAddonDll::CheckAPIVersion(int type)
+
+  if (strcasecmp(settingName, "__addonpath__") == 0)
+  {
+    strcpy((char*) settingValue, addon->Path().c_str());
```
don't use strcpy, so easy to mess up and write random memory.
get_Setting need to allocate the string here and return it or know about the size but preferably allocate it.

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -425,5 +523,233 @@ bool CAddonDll::CheckAPIVersion(int type)
+            type == "programs" || type == "fileenum"  ||
+            type == "file"     || type == "labelenum")
+        {
+          strcpy((char*) settingValue, addon->GetSetting(id).c_str());
```
same as above

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -425,5 +523,233 @@ bool CAddonDll::CheckAPIVersion(int type)
+        }
+        else if (type == "number" || type == "enum")
+        {
+          *(int*) settingValue = (int) atoi(addon->GetSetting(id).c_str());
```
use c++ style casts

ksooo replied:
ALthough you are right, please don't forget that larger parts of this code are just moved from elsewhere during refactoring. Question is how far to go - means is fixing all sh*t contained in the original code in scope?

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -425,5 +523,233 @@ bool CAddonDll::CheckAPIVersion(int type)
+        }
+        else if (type == "bool")
+        {
+          *(bool*) settingValue = (bool) (addon->GetSetting(id) == "true" ? true : false);
```
use c++ style casts

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -425,5 +523,233 @@ bool CAddonDll::CheckAPIVersion(int type)
+          const char *option = setting->Attribute("option");
+          if (option && strcmpi(option, "int") == 0)
+          {
+            *(int*) settingValue = (int) atoi(addon->GetSetting(id).c_str());
```
use c++ style casts

Paxxi commented on xbmc/addons/AddonDll.cpp:
```diff
@@ -425,5 +523,233 @@ bool CAddonDll::CheckAPIVersion(int type)
+          }
+          else
+          {
+            *(float*) settingValue = (float) atof(addon->GetSetting(id).c_str());
```
use c++ style casts

AlwinEsch replied:
Want to leave them currently like it is, also thought to change them but with the coming https://github.com/xbmc/xbmc/pull/12125 I want to hold differences to them on "xbmc/addons/interfaces/Addon/AddonCallbacksAddon.cpp" so small as possible.

OK for you?

Paxxi replied:
if it's planned for a future PR then it's OK for me.

Paxxi reviewed (changes requested):
sorry to be so harsh but these things need to be addressed.

Keep up the good work though, the overall design looks good.

AlwinEsch commented:
Currently is much only a c&p from the old way. Because of the size I want todo them step by step, otherwise it breaks the current addons who not attached.

AlwinEsch commented:
The good think with the c++ addon headers is that it can be fixed without bigger changes on addons, the "C" part is completely done from the headers.

ksooo commented:
@Paxxi all very valid points, but may I suggest to do the cleanup once all pieces are in place? Currently, larger parts are just temporary and will not be present once everything for the addon rewrite got merged, thus tweaking right now might be work for the garbage bin.

AlwinEsch commented:
jenkins build this please

AlwinEsch committed:
[addons] add interface callbacks to read addon settings

Some addons (also Screensaver) use own settings. With this becomes the
support added on new way to read them on addon.

Due to the big use of them is it added to the addon base interface and not
separated to another.

AlwinEsch commented:
jenkins build this please

Paxxi commented:
@ksooo I would agree to that in theory. In reality there's almost never anyone who goes back and cleans up newly merged code. Style issues I can let pass but the bits like leaking memory shouldn't be merged and forgotten about.

ksooo commented:
> the bits like leaking memory shouldn't be merged and forgotten about.

 @Paxxi agreed, but there is no memleak with <code>CAddonStatusHandler</code>. This class is derived from <code>CThread</code> and if constructed with parameter bSameProcess set to false, <code>CThread::Create(true)</code> will be called from inside <code>CAddonStatusHandler</code> ctor. "true" means "thread auto deletes itself" here. I agree that this is no good class design, but no memleak.

 What else do you see as blocker for this PR?

AlwinEsch committed:
[addons] bring in new addon creation way for multiple instances

This is the first big major part to bring in support about multiple
instances.

There is still not the multiple instance way possible but prepared for
them.

This change allow several addons to match coming style and have equal
with kodi-agile.

AlwinEsch commented:
@Paxxi think to have all the parts changed and fixed, thanks a lot! Looks now better and learned a new part (`= { 0 }` :D ).

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
OK to merge?

ksooo commented:
Go for it!

AlwinEsch committed:
[screensaver] change addon interface to the coming style

This request change screensaver to the complete C++ style of
related interface headers.

This style is also for multiple instances used, but not needed
on Screensaver. This request is mostly to see the new style.

The other thing is that the kodi-agile and master are equal
on addon side for screensavers.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Thanks a lot @MartijnKaijser
